File: vol2/ch22/LeetCode300.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n=nums.size();
        vector<int> ans={nums[0]};
        for (int i=1;i<n;i++) {
            int curd=nums[i];
            if (curd>ans.back())        //大于ans中最大数时直接添加
                ans.push_back(curd);
            else {
                auto it=lower_bound(ans.begin(),ans.end(),curd);
                *it=curd;               //替换
            }
        }
        return ans.size();
    }
};

File: vol2/ch22/LeetCode300.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n=len(nums)
        ans=[nums[0]]
        for i in range(1,n):
            curd=nums[i]
            if curd>ans[-1]:        #大于ans中最大数时直接添加
                ans.append(curd)
            else:
                j=bisect.bisect_left(ans,curd)
                ans[j]=curd         #替换
        return len(ans)
